compose/ui/ui/src/wasmJsMain/kotlin/androidx/compose/ui/window/SystemThemeObserver.wasm.kt

```diff
@@ -19,13 +19,58 @@ package androidx.compose.ui.window
 import androidx.compose.runtime.State
 import androidx.compose.runtime.mutableStateOf
 import androidx.compose.ui.SystemTheme
+import org.w3c.dom.MediaQueryList
+import org.w3c.dom.MediaQueryListEvent
 import org.w3c.dom.Window
+import org.w3c.dom.events.Event
 
-internal actual fun getSystemThemeObserver(window: Window): SystemThemeObserver {
-    return object : SystemThemeObserver {
-        override val currentSystemTheme: State<SystemTheme>
-            get() = mutableStateOf(SystemTheme.Unknown)
+internal class SystemThemeObserverImpl(window : Window) : SystemThemeObserver {
 
-        override fun dispose() {}
+    override val currentSystemTheme: State<SystemTheme>
+        get() = _currentSystemTheme
+
+    private val media: MediaQueryList by lazy {
+        window.matchMedia("(prefers-color-scheme: dark)")
+    }
+
+    private val _currentSystemTheme = mutableStateOf(
+        when {
+            !isMatchMediaSupported() -> SystemTheme.Unknown
+            media.matches -> SystemTheme.Dark
+            else -> SystemTheme.Light
+        }
+    )
+
+    private val listener: (Event) -> Unit = { event ->
+        _currentSystemTheme.value = if ((event as MediaQueryListEvent).matches)
+            SystemTheme.Dark else SystemTheme.Light
+    }
+
+    override fun dispose() {
+        if (isMatchMediaSupported()){
+            try {
+                media.removeEventListener("change", listener)
+            } catch (t : Throwable){
+                media.removeListener(listener)
+            }
+        }
+    }
+
+    init {
+        if (isMatchMediaSupported()) {
+            try {
+                media.addEventListener("change", listener)
+            } catch (t: Throwable) {
+                media.addListener(listener)
+            }
+        }
     }
-}
\ No newline at end of file
+}
+
+internal actual fun getSystemThemeObserver(window: Window): SystemThemeObserver =
+    SystemThemeObserverImpl(window)
+
+// supported by all browsers since 2015
+// https://developer.mozilla.org/en-US/docs/Web/API/Window/matchMedia
+@JsFun("window.matchMedia != undefined")
```

Comment:
it crashes at runtime when running the demo:
```
./gradlew :compose:mpp:demo:wasmJsBrowserRun
```

This suggestion fixes it:

```suggestion
@JsFun("() => window.matchMedia != undefined")
```

Reply:
Applied it.

Reply:
Wondering why it didn't fail for you:
> Test: Built locally and used with a wasmJs compose app, and observed that dark mode was set correctly

Reply:
Hmm odd, is that a Kotlin error or a browser error?

Reply:
It appears in a browser

Reply:
I see, I get that error now as well. Maybe I was using an older version of a browser that didn't have that check?

Thank you for catching that :sweat_smile: